ports/nrf/modules/machine/pwm.c

```diff
@@ -37,94 +37,81 @@
 #include "genhdr/pins.h"
 #include "pwm.h"
 
-#if NRF52
+#if defined(NRF52_SERIES)
 // Use PWM hardware.
-#include "hal_pwm.h"
+#include "nrfx_pwm.h"
 #endif
 
-#ifdef MICROPY_HW_PWM0_NAME
-PWM_HandleTypeDef PWMHandle0 = {.instance = NULL};
+typedef enum {
+    MODE_LOW_HIGH,
+    MODE_HIGH_LOW
+} pwm_mode_t;
+
+typedef struct {
+    uint8_t        pwm_pin;
+    uint8_t        duty;
+    uint16_t       pulse_width;
+    uint16_t       period;
+    nrf_pwm_clk_t  freq;
+    pwm_mode_t mode;
+} machine_pwm_config_t;
+
+typedef struct _machine_hard_pwm_obj_t {
+    mp_obj_base_t          base;
+    const nrfx_pwm_t *     p_pwm;
+    machine_pwm_config_t * p_config;
+} machine_hard_pwm_obj_t;
+
+STATIC const nrfx_pwm_t machine_hard_pwm_instances[] = {
+#if NRF52
+    NRFX_PWM_INSTANCE(0),
+    NRFX_PWM_INSTANCE(1),
+    NRFX_PWM_INSTANCE(2),
+#elif NRF52840
+    NRFX_PWM_INSTANCE(3),
+#else
+    NULL
```

Comment:
It's a bit odd to put `NULL` here, but I guess since PWM isn't supported on the nrf51 anyway it doesn't really matter.